docs/internal/types.md

```diff
@@ -0,0 +1,287 @@
+  var common : I64;
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 2;
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class or interface and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
+
+They can also have a keyword `embed`, which means ...
+
+### Methods
+
+Methods are functions that belong to a class or interface which can have arguments and a return type.
+
+Methods can be called statically (via `Class::method()`) or dynamically (via `object.method()`).
+On the static case, the `self` pointer is a descriptor for the type, while on the dynamic case, it's a descriptor to the instance (and its vtable, type info, etc).
+
+### Constructors
+
+Verona distinguishes memory allocation from object construction.
+Verona's built-in `new` function fully constructs an object and so must be called with values for all of the fields in the object.
+It may be used only from static methods on the class and so is usually used by factory methods.
+By convention, factory methods are called `create()` and there is syntactic sugar to allow `T()` to implicitly call `T::create()`.
+
+Example:
+```ts
+class Capibara
+{
+  let name : String & imm;
+  var weight : F32; // Kgs
+
+  // Create a baby capibara
+  newborn(name : String) : Capibara
+  {
+    var self : Capibara = new Capibara(name, 3);
+    self; // Returns `self`
+  }
+
+  eat(self: Capibara, food : F32)
+  {
+    self.weight += food;
+  }
+
+  poop(self: Capibara)
+  {
+    self.weight -= 0.1; // Tiny poop!
+  }
+};
+```
+
+On the example above, the `newborn` method creates a `Capibara` with a specific name and a default weight.
+The `name` can't change, so we declare it with `let`, while the `weight` does change, so it's a `var`.
+
+There's nothing special about the method `newborn` in comparison with others, apart from the fact that it allocates, initialises and returns an object of the same type.
+
+## Destructors
+
+Destructors are still under debate, but they won't have to deallocate memory like in C++, just execute finalisation tasks like closing file handles, running some cleanup routines, checking final consistency, etc.
+
+## Primitive Types
+
+Verona treats numeric types specially in the compiler but their representation isn't special in the language.
+
+Verona's numeric types are singleton types (no fields) that have their machine representation based on their names.
+So, `I32` is a signed integer with 32-bit width (in LLVM's parlance, an `i32`) and so on.
+
+The full list of primitive numeric types are:
+ * Unsigned integers: `U8`, `U16`, `U32`, `U64` and `U128`
+ * Signed integers: `I8`, `I16`, `I32`, `I64` and `I128`
+ * Floating Point: `F32` and `F64`
+ * System types: `ULength` and `ILength` (machine-word sized integers)
```

Comment:
```suggestion
 * System types: `ULength` and `ILength` (size large enough to describe any offset into the largest array that it's possible to allocate)
```

I find 'machine word' confusing since Intel decided that word is 16 bits, a double-word is 32 and a quad word is 64.  Everyone seems to have a different definition of what a word means these days (when clearly it should be an arbitrary-length sequence of 6-bit EBCDIC characters).